// VERSION: 2
// ======================================================================== //
// BLOSPRAY - OSPRay as a Blender render engine                             //
// Paul Melis, SURFsara <paul.melis@surfsara.nl>                            //
// Protocol messages                                                        //
// ======================================================================== //
// Copyright 2018-2019 SURFsara                                             //
//                                                                          //
// Licensed under the Apache License, Version 2.0 (the "License");          //
// you may not use this file except in compliance with the License.         //
// You may obtain a copy of the License at                                  //
//                                                                          //
//     http://www.apache.org/licenses/LICENSE-2.0                           //
//                                                                          //
// Unless required by applicable law or agreed to in writing, software      //
// distributed under the License is distributed on an "AS IS" BASIS,        //
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. //
// See the License for the specific language governing permissions and      //
// limitations under the License.                                           //
// ======================================================================== //

syntax = "proto3";

message ClientMessage
{
    enum Type {
        HELLO = 0;                      // XXX not used atm
        
        RETRIEVE_SCENE = 10;
        CLEAR_SCENE = 11;
        UPDATE_SCENE = 12;       
        
        UPDATE_OBJECT = 20;
        UPDATE_DATA = 21;

        DELETE_OBJECT = 30;
        DELETE_DATA = 31;
        
        START_RENDERING = 40;
        CANCEL_RENDERING = 41;
        
        QUERY_BOUND = 80;
        
        QUIT = 99;
    }
    
    Type    type = 1;
        
    //bool    bool_value = 10;
    uint32  uint_value = 20;             // HELLO     
    //float   float_value = 30;
    string  string_value = 40;          // QUERY_BOUND
}

message QueryBoundResult
{
    bool    success = 1;
    string  message = 2;
    uint32  result_size = 3;
}


message RenderResult 
{
    enum Type {
        FRAME = 0;
        CANCELED = 1;
        DONE = 2;
    }
    
    Type    type = 1;
    
    // FRAME
    uint32  sample = 2;
    string  file_name = 3;          // Only used for server-internal purposes
    uint32  file_size = 4;
    float   memory_usage = 5;
}

// Scene

message SceneElement
{   
    enum Type {
        NONE = 0;                   // Used for signaling last object was sent
        
        // Regular Blender mesh
        MESH_OBJECT = 10;
        MESH_DATA = 11;

        // OSPRay geometry 
        GEOMETRY_OBJECT = 20;
        GEOMETRY_DATA = 21;

        // OSPRay volume
        VOLUME_OBJECT = 30;
        VOLUME_DATA = 31;
        
        // OSPRay isosurface
        ISOSURFACE_OBJECT = 40;
        
        // OSPRay slices
        SLICES_OBJECT = 50;        
        
        // OSPRay scene (list of transformed Groups)
        SCENE_OBJECT = 60;
        SCENE_DATA = 61;
        
        // A single light
        LIGHT = 100;
    }
    
    Type            type = 1;
    string          name = 2;
    string          properties = 3;
    
    repeated float  object2world = 10;
    string          data_link = 11;
}

message MeshData
{
    enum Flags {
        NONE = 0;
        NORMALS = 1;
        VERTEX_COLORS = 2;
        // = 4;
        // = 8;
    }

    uint32          flags = 1;
    uint32          num_vertices = 10;
    uint32          num_triangles = 11;
}

// Settings

message ImageSettings
{
    uint32          width = 1;
    uint32          height = 2;
    repeated float  border = 3;
}

message CameraSettings
{
    enum Type {
        PERSPECTIVE = 0;
        ORTHOGRAPHIC = 1;
        PANORAMIC = 2;
    }
    
    Type    type = 1;
    string  object_name = 2;
    string  camera_name = 3;

    repeated float position = 10;
    repeated float view_dir = 11;
    repeated float up_dir = 12;
    
    // Perspective
    float fov_y = 20;                // Degrees
    
    // Orthographic
    float height = 30;
    
    // Perspective, orthographic
    float aspect = 40;
    
    // Common
    float clip_start = 50;
    
    float dof_focus_distance = 60;
    float dof_aperture = 61;
}

message RenderSettings
{
    string          renderer = 1;
    
    uint32          samples = 2;
    uint32          ao_samples = 3;
    
    repeated float  background_color = 4;   // RGBA
    
    bool            shadows_enabled = 5;    // XXX no longer in API2?
}

message Light 
{
    enum Type {
        POINT = 0;
        SUN = 1;
        SPOT = 2;
        AREA = 3;
    }
    
    Type            type = 1;
    repeated float  object2world = 2;
    string          object_name = 3;
    string          light_name = 4;
    
    
    // Common values
    repeated float  color = 10;
    float           intensity = 11;
    bool            visible = 12;
    
    // Per-type values
    repeated float  direction = 20;         // Directional, spot
    float           angular_diameter = 21;  // Directional
    repeated float  position = 22;          // Point/sphere, spot, quad
    float           radius = 23;            // Point/sphere, spot
    float           opening_angle = 24;     // Spot
    float           penumbra_angle = 25;    // Spot
    repeated float  edge1 = 26;             // Area/Quad
    repeated float  edge2 = 27;             // Area/Quad
}   

message LightSettings
{
    float           ambient_intensity = 1;
    repeated float  ambient_color = 2;
    
    repeated Light  lights = 3;
}

// Plugins

// Returned by generate function
message GenerateFunctionResult
{
    bool            success = 1;
    string          message = 2;        // XXX use JSON for returning multiple lines/messages?
    
    string          hash = 3;
}

